src/math/dct.c

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/audio/format.h>
+#include <sof/math/matrix.h>
+#include <sof/math/dct.h>
+#include <sof/math/sqrt.h>
+#include <sof/math/trig.h>
+#include <errno.h>
+#include <stdint.h>
+
+#define PI_Q29 Q_CONVERT_FLOAT(3.1415926536, 29)
+#define TWO_PI_Q24 Q_CONVERT_FLOAT(6.2831853072, 24)
+#define ONE_OVER_SQRT_TWO Q_CONVERT_FLOAT(0.7071067812, 31)
+#define HALF_Q1 Q_CONVERT_FLOAT(0.5, 1)
+#define TWO_Q29 Q_CONVERT_FLOAT(2, 29)
+
+/* Return a matrix for discrete cosine transform (DCT). The DCT type is hard
+ * coded to DCT-II with orthogonal matrix. See
+ * https://en.wikipedia.org/wiki/Discrete_cosine_transform#DCT-II
+ *
+ * The size is (num_in, num_out). Currently there are not other input parameters.
+ * TODO: Add option for no ortho norm.
+ */
+
+/**
+ * \brief Initialize a 16 bit DCT matrix. The actual DCT transform is a matrix
+ * multiply with the returned matrix.
+ * \param[in,out]  dct  In input provide DCT type and size, in output the DCT matrix
+ */
+int dct_initialize_16(struct dct_plan_16 *dct)
+{
+	int16_t dct_val;
+	int32_t arg;
+	int32_t cos;
+	int32_t c1;
+	int16_t c2;
+	int16_t nk;
+	int n;
+	int m;
```

Comment:
Unsigned int are dangerous in arithmetic. Trust me that I have had nasty issues in mass market devices with those, so I want to avoid them when possible. Comparing numbers always needs subtraction that is a signed operation so the unsigned code becomes signed any way when executed in CPU ALU. Despite more compact looking source code sometimes with unsigned, there's compiler generated assembler that does the work. There's another way to prevent overflow when using signed type. And in this case the matrix sizes are much smaller than int16 range.

Quotation from inventor of C++ Donald Knuth:
![Screenshot from 2022-08-17 19-11-25](https://user-images.githubusercontent.com/28754793/188839008-18d2ecb8-d7ee-4bd4-97b4-d7c1987de121.png)

Reply:
I realized that the code needs a check for 

```
	if (dct->num_in < 1 || dct->num_out < 1)
		return -EINVAL;

```

So, use of int types does not even create more C code lines than unsigned. OK?

Reply:
What I meant before: you must use `unsigned` if you INTEND to overflow. I did NOT mean use `unsigned` "just in case something unexpected happens and it overflows". If "something unexpected happens" then you are _already screwed_ anyway; don't even think about signs and overflows.

> Unsigned int are dangerous in arithmetic.

_Mixing different types_ is very dangerous because then no one can ever remember what gets implicitly converted to what. Integer conversions are an absolute nightmare https://stackoverflow.com/questions/46073295/implicit-type-promotion-rules

However I think using `unsigned int` _consistently everywhere_ (if/when that makes sense) is safe, please prove me wrong with a counter example. Should be easy to find on the Internet if there is one :-)

> Comparing numbers always needs subtraction that is a signed operation so the unsigned code becomes signed any way when executed in CPU ALU.

This is a shortcut / simplification. Here's one particular step out of the way too many steps at: https://stackoverflow.com/questions/46073295/implicit-type-promotion-rules
> Otherwise, if the operand that has unsigned integer type has rank greater or equal to the rank of the type of the other operand, then the operand with signed integer type is converted to the type of the operand with **unsigned** integer type.


I think one of the main issues is C has signed data and agnostic operators whereas machine code has agnostic data and signed operators. 
https://en.wikibooks.org/wiki/X86_Assembly/Control_Flow
Also, C has tried to pretend it is portable AND "low-level" and has failed miserably https://faultlore.com/blah/c-isnt-a-language/

Reply:
> inventor of C++ Donald Knuth

@singalsu forgive me, but I think it was [Bjarne Stroustrup](https://en.wikipedia.org/wiki/Bjarne_Stroustrup) ;-)

Reply:
> Comparing numbers always needs subtraction that is a signed operation

@singalsu I also don't think this is true: I think at least on some arches there are both signed and unsigned compare assembly instructions. And C explicitly produces different results for signed and unsigned comparisons. `-1 < 1` when compared as unsigned, but the opposite is true when compared as signed

Reply:
@marc-hb Thanks for deep insight! Though the examples make me even more paranoid about unsigned types. Really need to understand the rules.

Reply:
don't think either of them is intrinsically better or worse. I just know that I used `int` almost exclusively before, but then I started getting some errors or warnings - both from the compiler and at run-time. So for me personally it proved to be "safer" to use `unsigned int` where I think that I know that it won't cause problems :-) The simplest example is `for`-loops with guaranteed non-negative limits. Then also some size or "quantity" variables, that also should never be negative. Every time you use `size_t` you use an unsigned type. Or think about all uses of `uint32_t` in SOF - those are all `unsigned int` too. I agree, that type conversion / promotion rules might not seem very clear or intuitive. So care should be applied. Especially with comparisons.

Reply:
Makes sense. Though in my view size is bit like money, spending too much doesn't suddenly turn into big wealth, instead I get debt. I think it's a matter of programming style what everyone is comfortable with and knows how to avoid issues. Also in for loop exceeding positive range of int is probably not a good idea so int in 100% my cases is sufficient for for loops.

I'm not in C relying in some assumed overflow behavior with signed integers (except a norm function that supported negative.. it got a fix though it never was demonstrated to fail). Though I have done it in HW where  the behavior is known. It's a shame that C standard didn't mandate 2s complement for ints. I've never seen a C platform where int wouldn't be 2s complement. Sign-magnitude notation has a bigger HW gate count in HW and the benefits for accuracy are questionable. Instead there's duplicated zero to take care of.

Reply:
> It's a shame that C standard didn't mandate 2s complement for ints. I've never seen a C platform where int wouldn't be 2s complement.

Even when signed int is 2-complement, overflow is STILL undefined. Proof:
- https://github.com/zephyrproject-rtos/zephyr/pull/41553

"undefined" == the compiler's craziness will always surprise you and may cost you many days of work.

"implementation defined" == works now but it's a time-bomb.


> So for me personally it proved to be "safer" to use unsigned int where I think that I know that it won't cause problems :-) 

The important point that Stroustrup seems to miss: unsigned ints are better defined and more portable than signed ints.

I'm fairly sure ALL problems are either overflow- or conversion- related. Areas where none happens should be safe with either signed or unsigned. But of course it's impossible to avoid all conversions and they must be extremely well scrutinized.